docs/decisions/0017-reimplement-asset-processing.rst

```diff
@@ -0,0 +1,268 @@
+     
+   * - + **Build stage 1: Copy npm-installed assets** from node_modules to other folders in edx-platform. They are used by certain especially-old legacy LMS & CMS frontends that are not set up to work with npm directly.
+
+     - ``paver update_assets --skip-collect``
+
+       Implemented in Python within update_assets. There is no standalone command for it.
+
+     - ``assets/build.sh npm``
+
+       Pure Bash reimplementation.
+ 
+   * - + **Build stage 2: Copy XModule fragments** from the xmodule source tree over to input directories for Webpack and SCSS compilation. This is required for a hard-coded list of old XModule-style XBlocks. This is not required for new pure XBlocks, which include (or pip-install) their assets into edx-platform as ready-to-serve JS/CSS/etc fragments.
+
+     - ``paver process_xmodule_assets``, or
+
+       ``xmodule_assets``
+
+       Equivalent paver task and console script, both pointing at to an application-level Python module. That module inspects attributes from legacy XModule-style XBlock classes in order to determine which static assets to copy and what to name them.
+
+     - ``assets/build.sh xmodule``
+
+       A Bash implementation of XModule asset copying. The aforementioned attributes will be moved from the XModule-style XBlock classes into a simple static JSON file, which the Bash script will be able to read.
+       
+       The initial implementation of build.sh may just point at ``xmodule_assets``.
+   
+   * - + **Build stage 3: Run Webpack** in order to to shim, minify, otherwise process, and bundle JS modules. This requires a call to the npm-installed ``webpack`` binary.
+
+     - ``paver webpack``
+
+       Python wrapper around a call to webpack. Invokes the ``./manage.py [lms|cms] print_setting`` multiple times in order to determine Django settings, adding which can add 20+ seconds to the build.
+
+     - ``assets/build.sh webpack``
+
+       Bash wrapper around a call to webpack. The script will accept parameters for Django settings rather than looking them up. Open edX distributions, such as Tutor, can choose how to supply the Django-setting-derived parameters in an efficient manner.
+   
+   * - + **Build stage 4: Compile default SCSS** into CSS for legacy LMS/CMS frontends.
+
+     - ``paver compile_sass``
+
+       Paver task that invokes ``sass.compile`` (from the libsass Python package) and ``rtlcss`` (installed by npm) for several different directories of SCSS.
+
+       Note: libsass is pinned to a 2015 version with a non-trivial upgrade path. Installing it requires compiling a large C extension, noticeably affecting Docker image build time.
+
+     - ``assets/build.sh common``
```

Comment:
Naming this command `common` seems a bit odd to me - are we assuming stuff other than scss gets compiled here in the future?  Maybe this is just `css` and we add a `common` later if we need it, as a higher level command that calls `css` and whatever-other-things-feel-common?

Reply:
I chose `common` to ease the migation for Tutor (which uses `openedx-assets common`) but I do like `css` more.

I'll think on this!

Reply:
+1 to explicit naming here, though I understand it's a tradeoff with ease of transition.

Reply:
FWIW it's perfectly fine to modify the existing openedx-assets script in Tutor, including to rename some of the commands.